File: nafos-core/src/main/java/nafos/bootStrap/handle/http/NsRespone.java
package nafos.bootStrap.handle.http;

import io.netty.handler.codec.http.DefaultFullHttpResponse;
import io.netty.handler.codec.http.HttpHeaderNames;
import io.netty.handler.codec.http.HttpResponseStatus;
import io.netty.handler.codec.http.HttpVersion;
import io.netty.handler.codec.http.cookie.Cookie;
import io.netty.handler.codec.http.cookie.DefaultCookie;
import io.netty.util.AsciiString;
import nafos.core.util.AESUtil;
import nafos.core.util.SnowflakeIdWorker;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.List;

/**
 * @author 作者 huangxinyu
 * @version 创建时间：2018年1月25日 下午5:52:20
 * 类说明
 */
public class NsRespone extends DefaultFullHttpResponse {

    /**
     * cookie
     */
    private List<Cookie> cookies = new ArrayList<Cookie>();

    public NsRespone(HttpVersion version, HttpResponseStatus status) {
        super(version, status);
    }

    public NsRespone setCookie(String key, String value) {
        cookies.add(new DefaultCookie(key, value));
        return this;
    }

    public NsRespone setNafosCookieId(String value) {
        cookies.add(new DefaultCookie("nafosCookie", URLEncoder.encode(AESUtil.encrypt(value))));
        return this;
    }

    public List<Cookie> getCookies() {
        return cookies;
    }

    public void setCookies(List<Cookie> cookies) {
        this.cookies = cookies;
    }

    public void setHeader(String key, String value) {
        headers().set(key, value);
    }

    /**
     * @Desc 返回文件
     * @Param filePath  要返回的文件路径
     * @Param fileName  返回文件在前端的名字显示
     * @Author hxy
     * @Time 2019/10/24 12:19
     */
    public NsRespone returnFile(String filePath, String fileName) {
        File file = new File(filePath);
        this.setHeader("Content-Disposition", "attachment;fileName=" + fileName);
        this.setHeader(HttpHeaderNames.CONTENT_TYPE.toString(), "multipart/form-data");
        this.setStatus(HttpResponseStatus.OK);
        FileInputStream in = null;
        try {
            in = new FileInputStream(file);
            this.content().writeBytes(in.getChannel(), 0L, in.available());
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (in != null) {
                try {
                    in.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return this;
    }

    /**
     * @Desc     重定向
     * @Author   hxy
     * @Time     2019/10/24 13:07
     */
    public NsRespone returnRedirect(String path){
        this.setHeader("location",path);
        this.setStatus(HttpResponseStatus.FOUND);
        return this;
    }
}

File: server/src/main/kotlin/nafos/server/handle/http/NsRespone.kt
package nafos.server.handle.http

import io.netty.handler.codec.http.DefaultFullHttpResponse
import io.netty.handler.codec.http.HttpHeaderNames
import io.netty.handler.codec.http.HttpResponseStatus
import io.netty.handler.codec.http.HttpVersion
import io.netty.handler.codec.http.cookie.Cookie
import io.netty.handler.codec.http.cookie.DefaultCookie
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException
import java.util.ArrayList

/**
 * @Description respone
 * @Author      xinyu.huang
 * @Time        2019/11/24 21:02
 */
class NsRespone(version: HttpVersion, status: HttpResponseStatus) : DefaultFullHttpResponse(version, status) {
    var cookies: MutableList<Cookie> = ArrayList()

    fun setCookie(key: String, value: String): NsRespone {
        cookies.add(DefaultCookie(key, value))
        return this
    }

    fun setHeader(key: String, value: String) {
        headers().set(key, value)
    }

    /**
     * @Desc 返回文件
     * @Param filePath  要返回的文件路径
     * @Param fileName  返回文件在前端的名字显示
     * @Author hxy
     * @Time 2019/10/24 12:19
     */
    fun returnFile(filePath: String, fileName: String): NsRespone {
        val file = File(filePath)
        this.setHeader("Content-Disposition", "attachment;fileName=$fileName")
        this.setHeader(HttpHeaderNames.CONTENT_TYPE.toString(), "multipart/form-data")
        this.status = HttpResponseStatus.OK
        var `in`: FileInputStream? = null
        try {
            `in` = FileInputStream(file)
            this.content().writeBytes(`in`.channel, 0L, `in`.available())
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            if (`in` != null) {
                try {
                    `in`.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }
        return this
    }

    fun returnFile(byteArray: ByteArray, fileName: String): NsRespone {
        this.setHeader("Content-Disposition", "attachment;fileName=$fileName")
        this.setHeader(HttpHeaderNames.CONTENT_TYPE.toString(), "multipart/form-data")
        this.status = HttpResponseStatus.OK
        this.content().writeBytes(byteArray)
        return this
    }

    /**
     * @Desc     重定向
     * @Author   hxy
     * @Time     2019/10/24 13:07
     */
    fun returnRedirect(path: String): NsRespone {
        this.setHeader("location", path)
        this.status = HttpResponseStatus.FOUND
        return this
    }
}